Implemented interactive graphs using dygraphs

crleblanc opened the pull request:
Lots of modifications, now including our own js file and using a proxy
for the csv endpoints from mtr-api.  Investigated using protobufs from
javascript but it was bringing in a lot of dependencies so opted for CSV.
These graphs are available on all pages currently displaying SVG plots.

This is a large PR and I expect a fair bit of discussion.  I'm wondering what everyone's thoughts are on everything but especially:

* The use of the Interactive page struct element to determine if we're using interactive graphs
* The way I'm calling the JS code.  I'm new to JS so I could have missed a much better approach, possibly jquery.
* The way I'm getting the CSV data from the mtr-api.  I'm using a local proxy and XMLHttpRequest to get the data (async).  @sorennh mentioned jquery but my approach was to get it working first then improve it.
* The Dygraphs are using the "resolution" parameter because some of the data were getting pretty large and load times were unacceptable.  There are probably many different ways of solving this, any ideas are welcome.
* The buttons for 12 hours / 48 hours / 28 days / etc. look pretty ugly right now but I'm hoping we can replace this layout with something more pleasing.  The templates for these buttons are getting a bit large too.

I realise this isn't using the new weft.toml syntax but hopefully I can get it merged before @gclitheroe makes those changes!

junghao commented on mtr-ui/assets/tmpl/components.html:
```diff
@@ -154,15 +154,50 @@
+                if (!res) {
+                    res = "minute"
+                }
+
```
This default value checking doesn't seem necessary if you remove the p.Interactive here:
https://github.com/GeoNet/mtr/pull/213/files#diff-9af2073d500b90b22b1b90404f86a9e1R139

crleblanc replied:
Oh yes, good point.  When I started this branch I assumed we could use high resolution data with the interactive graphs and wouldn't require the "resolution" option.  However, this was too much data to transfer (as csv) and was causing long load times so I added the resolution param back in.

@gclitheroe mentioned that we should support "progressive loading" which would again eliminate the need for the resolution param since it would automatically fetch data at the correct resolution.  However, the SVG graphs still use the resolution param.

I wonder if we should simplify this PR as you suggest and do the progressive load in another branch?

junghao commented on mtr-ui/assets/js/graph.js:
```diff
@@ -0,0 +1,71 @@
+    request.setRequestHeader ("Accept", "text/csv");
+
+    request.onload = function() {
+        if (request.status == 200 || request.status == 404) {
```
I prefer to add a label (e.g. "NO DATA") on the plot when there's no data available. While I was debugging other bugs I found the "NO DATA" would be useful to tell if it's a bug from mtr-ui or mtr-api, or no data available( no data received).

junghao commented:
1. There's an input box at bottom-left corner to change the Y range but there's no label/description about it.
2. Is it a good idea to draw points in different colour when values are bad?
3. Will protobuf for javascript benefits the loading time when there's a lot of data? (But the browser has to decode it so I'm not sure it'll be better)

crleblanc commented:
Thanks @junghao, very good points.

1. The input box is a dygraphs option and I haven't seen a way to label it.  I thought the same thing, it would be nice if it had a label.
2. I haven't heard this raised before so hadn't implemented it.  It's probably straightforward if the users think it's a good idea.  We already display the thresholds as a grey box (similar to the SVGs) so I wonder if this is enough?
3. I was looking into protobufs before using CSVs.  They're a bit painful to use from JS because they bring in dependencies on Closure or CommonJS and others.  I discussed this with @gclitheroe and we both dislike the additional dependencies.  I'm sure it would speed things up though, CSV is pretty fat and parsing it in JS does take time.  However, if we include progressive loading this should make the graphs quite fast.

I talked to @andrewcowie a few minutes ago.  Apparently Librato takes a while to load it's graphs so they're ok if it takes a second or two to load.

bpeng commented on mtr-ui/assets/js/graph.js:
```diff
@@ -0,0 +1,70 @@
+var msOffset = new Date().getTimezoneOffset()*60*1000; // offset in milliseconds
+
+function utcToLocal(x) {
```
This may not work with NZDT, and the timezone offset should be based on the time of the data instead of the current time.

crleblanc replied:
Why wouldn't it work with NZDT?

The dates had been displayed in UTC but I can imagine this would be confusing to users.  Unlike the SVG plots it's not plotted in terms of time before present.  Dygraphs plots dates such as "2016/05/21 01:50:31" which would require the user to calculate the offset from UTC to local time.

Andrew wasn't here so I talked to Sam about it.  He says Librato display in NZDT by default and have a toggle to switch to UTC.  We could probably implement that but @gclitheroe suggested we stick with UTC for now and add this toggle later if requested.

I'll remove the local time option and rename the X axis label to "UTC Time" and update this PR.

bpeng replied:
OK, there used to be an issue with NZDT, but now it seems to return the correct offset on my browser.

msOffset value should be calculated for each data point as the data time could be in NZST while the current time is in NZDT.

bpeng commented on mtr-ui/assets/tmpl/app_plot.html:
```diff
@@ -4,6 +4,109 @@
+                    res = "minute"
+                }
+
+                var counterOptions = {
```
would suggest moving most of these code into the js file to make the page tidy, or use ```jQuery( document ).ready ```

crleblanc replied:
Thanks, I've moved that into the graphing code.

bpeng commented on mtr-ui/assets/js/graph.js:
```diff
@@ -0,0 +1,70 @@
+
+    request.onerror = function() {
+        // There was a connection error of some sort
+        throw "error downloading CSV data";
```
may need handle in the plot in stead of throwing the error

crleblanc replied:
I've modified the code as you suggest.  It now displays an empty plot with a title that contains detail  of the error message.  This will require a PR for mtr-api to improve error handling for empty CSVs.

bpeng commented:
Works well and code look fine, the csv format make the plot processing simple. 

Page layout need more consideration for switching between the static and interactive plots, but it is OK at this stage.

crleblanc commented:
Thanks @junghao and @bpeng, I think I've addressed your points and this PR should be ready to merge.

I agree the layout for switching between resolutions is cumbersome.  I'm looking at progressive updates which should solve this problem (in a separate PR).

crleblanc commented:
resolves GeoNet/mtr#170
